modyn/selector/internal/selector_strategies/presampling_strategies/label_balanced_presampling_strategy.py

```diff
@@ -0,0 +1,141 @@
+from typing import Optional
+
+from modyn.metadata_database.metadata_database_connection import MetadataDatabaseConnection
+from modyn.metadata_database.models import SelectorStateMetadata
+from modyn.selector.internal.selector_strategies.presampling_strategies import AbstractPresamplingStrategy
+from sqlalchemy import Select, asc, func, select
+from sqlalchemy.sql.expression import func as sql_func
+
+
+def get_fair_share(capacity: int, requests: list[int]) -> int:
```

Comment:
I think it's a cool way to compute how many samples should each class get. Thanks CCA :)

Reply:
hehe. I saw you wrote tests for this but did not fully check whether all potential cases are tested. if you are sure we covered everything great :)

Reply:
(I actually need to learn more about this algorithm. I would have thought if you say it should be balanced, then you'd just check how many datapoints you have of each class, and then return as many from each class as you can to sustain the fair share. Like you have 10k points, 10% presampling, classes A, B, C with 400, 400, and 300 samples, you would return 300, 300, 300 and go even below the presampling ratio, but maybe this algorithm is a better solution to this problem

Reply:
I copied the implementation from an ETH course and my exercise session. I hope it's correct ahah

Reply:
The solution you presented is the simple one, where you set the per_class size to the smallest class size (there's a query for that). 
My solution is used if we want exactly to match the target size (so returning 1000 samples in your example) and it's used to re-allocate the remaining samples. So, in your allocation, A and B would get 350 samples while C just 300 samples. The cool part of this solution is that you can just use LIMIT fair_share in the SQL query (since class C is below so it's unaffected by the LIMIT). There can be other solutions to reallocate the remaining samples but probably more complex on the SLQ side (like a limit different for each class)